services/ui_backend_service/api/run.py

```diff
@@ -93,13 +93,41 @@ async def get_all_runs(self, request):
                   $ref: '#/definitions/ResponsesError405'
         """
 
+        allowed_order = self._async_table.keys + ["user", "run", "finished_at", "duration", "status"]
+        allowed_group = self._async_table.keys + ["user"]
+        allowed_filters = self._async_table.keys + ["user", "run", "finished_at", "duration", "status"]
+
+        # JSONB tag filters combined with `ORDER BY` causes performance impact
+        # due to lack of pg statistics on JSONB fields. To battle this, first execute
+        # subquery of ordered list from runs_v3 table in and filter by tags on outer query.
+        # This needs more research in the future to further improve performance.
+        builtin_conditions, _ = builtin_conditions_query(request)
+        has_tag_filter = len([s for s in builtin_conditions if 'tags||system_tags' in s]) > 0
+        if has_tag_filter:
+            _, _, _, order, _, _ = pagination_query(request, allowed_order=allowed_order, allowed_group=[])
+            if len(order) > 0:
+                overwrite_table_name = "(SELECT * FROM runs_v3 {order_by}) AS runs_v3".format(
```

Comment:
might be just me, but `overwrite_table_name` doesn't clearly convey that this overrides the `FROM` in the produced query, maybe something like `overwrite_select_from` ? Not a blocker for merging.